src/jdom/devicehost.ts

```diff
@@ -0,0 +1,132 @@
+import { JDBus } from "./bus";
+import JDServiceHost from "./servicehost";
+import Packet from "./packet";
+import { shortDeviceId } from "./pretty";
+import { anyRandomUint32, toHex } from "./utils";
+import ControlServiceHost from "./controlservicehost";
+import { JDEventSource } from "./eventsource";
+import { JD_SERVICE_INDEX_CRC_ACK, PACKET_PROCESS, PACKET_SEND, SELF_ANNOUNCE } from "./constants";
+
+export default class JDDeviceHost extends JDEventSource {
+    private _bus: JDBus;
+    private readonly _services: JDServiceHost[];
+    public readonly deviceId: string;
+    public readonly shortId: string;
+
+    constructor(services: JDServiceHost[], options?: {
+        deviceId?: string;
+    }) {
+        super();
+        this._services = [new ControlServiceHost(), ...services];
+        this.deviceId = options?.deviceId;
+        if (!this.deviceId) {
+            const devId = anyRandomUint32(8);
+            for (let i = 0; i < 8; ++i) devId[i] &= 0xff;
+            this.deviceId = toHex(devId);
+        }
+        this.shortId = shortDeviceId(this.deviceId);
+        this._services.forEach((srv, i) => {
+            srv.device = this;
+            srv.serviceIndex = i;
+        });
+
+        this.handleSelfAnnounce = this.handleSelfAnnounce.bind(this);
+        this.handlePacket = this.handlePacket.bind(this);
+    }
+
+    protected log(msg: any) {
+        console.log(`${this.shortId}: ${msg}`);
+    }
+
+    get bus() {
+        return this._bus;
+    }
+
+    set bus(value: JDBus) {
+        if (value !== this._bus) {
+            this.stop();
+            this._bus = value;
+            this.start();
+        }
+    }
+
+    private start() {
+        if (!this._bus) return;
+
+        this._bus.on(SELF_ANNOUNCE, this.handleSelfAnnounce);
+        if (this._services.length)
+            this._bus.on([PACKET_PROCESS, PACKET_SEND], this.handlePacket)
+        this.log(`start host`)
+    }
+
+    private stop() {
+        if (!this._bus) return;
+
+        this._bus.off(SELF_ANNOUNCE, this.handleSelfAnnounce);
+        this._bus.off([PACKET_PROCESS, PACKET_SEND], this.handlePacket)
+        this.log(`stop host`)
+        this._bus = undefined;
+    }
+
+    private handleSelfAnnounce() {
+        const ctrl = this._services[0] as ControlServiceHost;
+        ctrl.announce();
+    }
+
+    services(): JDServiceHost[] {
+        return this._services.slice(0);
+    }
+
+    toString() {
+        return `host ${this.shortId}`;
+    }
+
+    async sendPacketAsync(pkt: Packet) {
+        if (!this._bus) return;
+
+        pkt.deviceIdentifier = this.deviceId;
+        // compute crc and send
+        pkt.sendCoreAsync(this.bus);
+        // send to current bus
+        this.bus.processPacket(pkt);
+    }
+
+    private handlePacket(pkt: Packet) {
+        const devIdMatch = pkt.deviceIdentifier == this.deviceId;
+        if (pkt.requiresAck && devIdMatch) {
+            pkt.requiresAck = false // make sure we only do it once
```

Comment:
not sure it makes sense here - in the makecode impl it clears the flag that is on the frame, not packet, but here we're probably not sharing memory